ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/precompiles/privacy/OnChainPrivacyPrecompiledContract.java

```diff
@@ -0,0 +1,466 @@
+    // call to affect the state
+    // privateTransactionProcessor.processTransaction(...) commits the state if the process was
+    // successful before it returns
+    final PrivateTransactionProcessor.Result getVersionResult =
+        checkCanExecute(
+            messageFrame,
+            currentBlockHeader,
+            publicWorldState,
+            privacyGroupId,
+            currentBlockchain,
+            disposablePrivateState,
+            privateWorldStateUpdater,
+            OnChainGroupManagement.GET_VERSION_METHOD_SIGNATURE);
+
+    if (version.equals(getVersionResult.getOutput())) {
+      return true;
+    }
+    LOG.debug(
+        "Privacy Group {} version mismatch for commitment {}: expecting {} but got {}",
+        privacyGroupId.toBase64String(),
+        messageFrame.getTransactionHash(),
+        getVersionResult.getOutput(),
+        version);
+    return false;
+  }
+
+  protected void persistPrivateState(
+      final Hash commitmentHash,
+      final Hash currentBlockHash,
+      final PrivateTransaction privateTransaction,
+      final Bytes32 privacyGroupId,
+      final PrivacyGroupHeadBlockMap privacyGroupHeadBlockMap,
+      final MutableWorldState disposablePrivateState,
+      final WorldUpdater privateWorldStateUpdater,
+      final PrivateTransactionProcessor.Result result,
+      final Bytes payload) {
+
+    LOG.trace(
+        "Persisting private state {} for privacyGroup {}",
+        disposablePrivateState.rootHash(),
+        privacyGroupId);
+    privateWorldStateUpdater.commit();
+    disposablePrivateState.persist();
+
+    final PrivateStateStorage.Updater privateStateUpdater = privateStateStorage.updater();
+
+    updatePrivateBlockMetadata(
+        commitmentHash,
+        currentBlockHash,
+        privacyGroupId,
+        disposablePrivateState.rootHash(),
+        privateStateUpdater);
+
+    final Bytes32 txHash = keccak256(RLP.encode(privateTransaction::writeTo));
+
+    final int txStatus =
+        result.getStatus() == PrivateTransactionProcessor.Result.Status.SUCCESSFUL ? 1 : 0;
+
+    final PrivateTransactionReceipt privateTransactionReceipt =
+        new PrivateTransactionReceipt(
+            txStatus, result.getLogs(), result.getOutput(), result.getRevertReason());
+
+    privateStateUpdater.putTransactionReceipt(currentBlockHash, txHash, privateTransactionReceipt);
+
+    if (!privacyGroupHeadBlockMap.contains(Bytes32.wrap(privacyGroupId), currentBlockHash)) {
+      privacyGroupHeadBlockMap.put(Bytes32.wrap(privacyGroupId), currentBlockHash);
+      privateStateUpdater.putPrivacyGroupHeadBlockMap(
+          currentBlockHash, new PrivacyGroupHeadBlockMap(privacyGroupHeadBlockMap));
+    }
+
+    if (payload
+        .toHexString()
+        .startsWith(OnChainGroupManagement.REMOVE_PARTICIPANT_METHOD_SIGNATURE.toHexString())) {
+      privacyGroupHeadBlockMap.remove(Bytes32.wrap(privacyGroupId));
```

Comment:
Yes - this is fine - all the nodes that are *not* being removed should continue to track the group. The reason this is here is so you can do `add` -> `remove` -> `add` .

We should discuss this on a call if you are not happy because I exhausted all other options before doing this.